### PR TITLE
More precise return types

### PR DESCRIPTION
**Pull request by omgMath**
This PR adds more precise return types (e.g. `pd.DataFrame` instead of `object`)

(Plus my IDE seems to have killed some white space)

### CODE REVIEW

**Comment by pitchmuc** [comment]
Thanks Nicola
For the getProjects and getSegments I was thinking using something like this
```python
from typing import Union

def myfunction(self,*args)->Union[pd.DataFrame,dict]
```
I would also start to type the variable within the code in order to help understanding the script a bit better. 

such as 
```python
path:str = "/my-path"
```
What do you think ?

**Comment by omgMath** [comment]
> For the getProjects and getSegments I was thinking using something like [...]

Hm. At least according to the code data is either a `pd.DataFrame` or a list of dictionaries, so I would stick with my proposal. Or what would be the reason?

> I would also start to type the variable within the code in order to help understanding the script a bit better.

I can tell you a little about our use case: We use the type annotations for type checking (with `mypy`), not necessarily for better readability/understanding. `mypy` infers types in some cases, for example, when running `mypy` over
```python
def test_fn() -> None:
    s = 'test'
    s = 15
```
it reports `error: Incompatible types in assignment (expression has type "int", variable has type "str")`, i.e. it already infers the type of `s` to be `str`, and reports an error, because afterwards an `int` is assigned to it.

So in conclusion, for our case this is not necessary.

**Comment by pitchmuc** [comment]
for the `getProjects` and `getSegments`, you can chose the format of the return with the `format` parameter. 
https://github.com/pitchmuc/adobe_analytics_api_2.0/blob/9af20e304a4bc02103bfa16a58e9780c1773bdf3/aanalytics2/aanalytics2.py#L1326-L1338
hence the possibility to be a dict or a pd.DataFrame (so Union as per the documentation : 
> [Union type; Union[X, Y] means either X or Y](https://docs.python.org/3/library/typing.html#typing.Union)
 
 -----------
>  mypy infers types in some cases, for example, when running mypy over
This is awesome. :)  I wanted to use mypy for some time now but didn't find time to learn how to use it efficiently yet. 
That is why I started to use type annotation for the long run.

I do not know if I will be able to achieve this so I didn't mention it but, basically I wanted to add this for those 2 reasons.
* Increase readibility (so you can better understand the code & expectation of code for new comers)
* Increase reliability (with mypy you can check type consistency when creating pipelines)

It would be great if you could provide some mypy examples.

**Comment by omgMath** [comment]
It is already a `Union`, because we defined two types at the top of the file:

```python
JsonOrDataFrameType = Union[pd.DataFrame, dict]
JsonListOrDataFrameType = Union[pd.DataFrame, List[dict]]
```

`JsonListOrDataFrameType` is then used as return type for the `getProjects` and `getSegments`.

>  This is awesome. :) I wanted to use mypy for some time now but didn't find time to learn how to use it efficiently yet.
That is why I started to use type annotation for the long run.

Yes, it is :smile: 
In our setup we have a `mypy` command running before each push and therefore are not allowed to push if there are type errors detected.

I would add types only to parameters and return values, I wouldn't add them to all of the variables. I would prefer e.g. `foo = 'test'` over `foo: str = 'test'`, because it is kinda clear from the assignment already. And the result when running `mypy` is the same in both cases.

**Comment by pitchmuc** [comment]
Thanks again for cleaning up my code a bit.
I'll do the release on pypi in the next days.

**Comment by Ryuno-Ki** [comment]
>  In our setup we have a `mypy` command running before each push and therefore are not allowed to push if there are type errors detected.

I'm quite sure, you could have it run as GitHub Actions, so you cannot _merge PR_s with type errors …
